### ISSUE TITLE
sc_mem_secure_alloc/free leave unaligned allocations locked in memory

### ISSUE THREAD

**Issue by dgalling** [OpenSC/OpenSC#3267]
### Problem Description

Context: Fedora 41 / OpenSC 0.25.1 (but the affected code is completely unchanged on master).

`sc_mem_secure_alloc` and `sc_mem_secure_free` use the `mlock` and `munlock` syscalls to lock pages in memory (to mitigate the chances of secret data being written to swap, I assume). Because `mlock` and `munlock` operate at page granularity, `sc_mem_secure_alloc` makes an attempt to get an aligned block of memory from `calloc` by rounding up the allocation size to the system page size. I do not know if POSIX mandates that `calloc` *always* return a page-aligned address for a page-size allocation, but one situation where it empirically does not is inside the Chromium browser, which uses its own PartitionAlloc allocator instead of glibc's.

`sc_mem_secure_alloc` behaves badly in a couple ways when `calloc` returns a pointer is that isn't page-aligned:

1. Any data on the page at lower addresses than what `calloc` returned gets `mlock`ed along with the intended buffer (not terrible, but probably not what you intended)
2. Because `sc_mem_secure_alloc` passes the rounded-up-to-page-size length to `mlock`, the page *after* the intended page also gets locked, even if the intended buffer does not touch that page
3. Because `sc_mem_secure_free` passes the *not* rounded-up-to-page-size length to `munlock`, the range of pages covered by the `munlock` call may be one less than the range of pages previously locked by `mlock`

I found it difficult to wrap my head around the issue at first, so here are some logs I added that helped me understand:

```
mlock(0x32d403f53600, 4096)
<<< PAGE TO BE LOCKED: 32d403f53000
<<< PAGE TO BE LOCKED: 32d403f54000
munlock((nil), 0)
mlock(0x32d402cfc800, 4096)
<<< PAGE TO BE LOCKED: 32d402cfc000
<<< PAGE TO BE LOCKED: 32d402cfd000
munlock(0x32d402cfc800, 307)
<<< PAGE TO BE UNLOCKED: 32d402cfc000
munlock(0x32d403f53600, 51)
<<< PAGE TO BE UNLOCKED: 32d403f53000
```

As you can see, the calls to `mlock` lock page N *and* N+1, but the calls to `munlock` only unlock page N. This actually causes Chromium to crash because it may later try to `madvise(MADV_DONTNEED)` the page that OpenSC left locked. Passing `madvise(MADV_DONTNEED)` a page that is currently locked in memory fails with EINVAL, and Chromium dies here.

### Proposed Resolution

`calloc` is probably not the right function to *guarantee* you get a whole page to yourself that you can `mlock`/`munlock` as you need for your secure alloc function. I recommend calling `mmap`/`VirtualAlloc` directly, bypassing the libc allocator.

From my reading of the man pages that ship on Fedora as well as MSVC's documentation, the alignment behavior that you're currently relying on is not necessarily guaranteed. Even if the alignment behavior is reasonably common, if what you need is your own page, bypassing the libc allocator is a less brittle way of doing that.

If calling `mmap`/`VirtualAlloc` sounds OK to you, let me know and I'll send a patch.

If that's unacceptable for some reason, I found that C11's `aligned_alloc` also worked around the issue. (If C11 is too new, there are older functions in POSIX that do the same thing).

### Steps to reproduce

The log snippet posted above was captured while using a smart card for TLS client authentication to a web service in a debug build of the Chromium browser from the tip of the master branch. I built OpenSC 0.25.1 from source (that is the version Fedora 41 ships), but as noted above, the implementation of `sc_mem_secure_alloc` and `sc_mem_secure_free` are unchanged on master.

**Comment by frankmorgner** [comment]
I have [tested aligned_alloc](https://github.com/frankmorgner/OpenSC/tree/aligned_alloc), but that function [doesn't seem to be available on Windows](https://ci.appveyor.com/project/frankmorgner/opensc-frankmorgner/builds/51051714), which leaves your suggested solution as only option, I guess.

However, [I think](https://github.com/OpenSC/OpenSC/pull/3269/commits/968dc08ee846b483b061fb516bf2d06e99d459a9#r1856388185) your original implementation needs to consider the page size when allocating the memory with VirtualAlloc.

**Comment by dgalling** [comment]
MSVC does have [_aligned_malloc](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/aligned-malloc?view=msvc-170) instead of the C11 standard function. Do note that the order of parameters is reversed from C11's `aligned_alloc` and that the documentation says you must use `_aligned_free` instead of `free` to free the region.

`_aligned_malloc` appears to be available going back to at least VS2015.

I [responded to your comment](https://github.com/OpenSC/OpenSC/pull/3269#discussion_r1856839572) about `VirtualAlloc` and page size on the PR.

**Comment by frankmorgner** [comment]
If we still require platform specific code with aligned_malloc, then it is of little use. I'll try to have a deeper look at the mmap documentation. If that looks good, I think we can adopt your initial proposal.

**Comment by frankmorgner** [comment]
Could you elaborate on the Chromium crashes? Do they happen regularly or under very specific circumstances? The question is whether we need a fix of the recent release.

**Comment by frankmorgner** [comment]
@Jakuje , @xhanulik, I think we need a fix-up release (0.26.1), once we have agreed on merging #3282 (first!) and #3281 (alternatively there is #3269 which has a workaround for the actual problem) - this seemed to have bugged Linux' chrome users for a year now...

**Comment by Jakuje** [comment]
I am trying to figure out what is different in chromium and in other browsers that load the OpenSC, such as Firefox, where I did not see this issue (but not that I would be using my yubikey that frequently).

Is it an issue that the chromium initializes OpenSSL with secure memory and OpenSC is reusing this context? I think we recently had some changes that will make the OpenSC to create its own context, but I do not know if this will change anything anything though.

**Comment by frankmorgner** [comment]
It is true that OpenSSL's secure memory is based on mlock as well (AFAIKT, it doesn't have support for Windows), but it is performing its initialization and management independent from OpenSC.

The issue we have here, is purely due to OpenSC's internal handling of mlock. So yes, this should also affect other software as well, but like you I have not experienced any problems thereof. Maybe Chromium is doing something under the hood that interacts with our usage of mlock and crashes the application as consequence...

**Comment by xhanulik** [comment]
> @Jakuje , @xhanulik, I think we need a fix-up release (0.26.1), once we have agreed on merging #3282 (first!) and #3281 (alternatively there is #3269 which has a workaround for the actual problem) - this seemed to have bugged Linux' chrome users for a year now...

@frankmorgner Sorry for delay. I think we can do the fix-up release - if it's okay, I can do it at the beginning of the January. Are there any other changes that should go there as well?

**Comment by frankmorgner** [comment]
That sounds good. I have created a new branch and cherry-picked what I think should be in a new fixup release.
